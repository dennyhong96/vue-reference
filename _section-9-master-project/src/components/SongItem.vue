<template>
  <li
    class="flex justify-between items-center p-3 pl-6 cursor-pointer transition 
          duration-300 hover:bg-gray-50"
  >
    <div>
      <router-link
        :to="{ name: 'Song', params: { id: song.id } }"
        class="font-bold block text-gray-600"
        >{{ song.modifiedName }}</router-link
      >
      <span class="text-gray-500 text-sm">{{ song.displayName }}</span>
    </div>

    <div class="text-gray-600 text-lg">
      <span class="comments">
        <i class="fa fa-comments text-gray-600"></i>
        {{ song.commentCount }}
      </span>
    </div>
  </li>
</template>

<script lang="ts">
import { defineComponent, PropType } from "vue";

import { SongWithId } from "@/types/Song";

export default defineComponent({
  name: "SongItem",

  props: {
    song: {
      required: true,
      type: Object as PropType<SongWithId>,
    },
  },
});
</script>
