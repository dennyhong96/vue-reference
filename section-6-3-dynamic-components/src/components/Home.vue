<template>
  <h1>Home</h1>
</template>

<script lang="ts">
import { defineComponent } from "vue";

export default defineComponent({
  mounted() {
    console.log("Home is mounted");
  },
  activated() {
    console.log("Home is activated");
  },
  unmounted() {
    console.log("Home is unmounted");
  },
  deactivated() {
    console.log("Home is deactivated");
  },
});
</script>
