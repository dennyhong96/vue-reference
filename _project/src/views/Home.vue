<template>
  <main>
    <!-- Introduction -->
    <section class="mb-8 py-20 text-white text-center relative">
      <div
        class="absolute inset-0 w-full h-full bg-contain introduction-bg"
        style="background-image: url(assets/img/header.png)"
      ></div>
      <div class="container mx-auto">
        <div class="text-white main-header-content">
          <!-- $t() is available after registering i18n as a plugin -->
          <h1 class="font-bold text-5xl mb-5">{{ $t("home.listen") }}</h1>
          <p class="w-full md:w-8/12 mx-auto">
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus
            et dolor mollis, congue augue non, venenatis elit. Nunc justo eros,
            suscipit ac aliquet imperdiet, venenatis et sapien. Duis sed magna
            pulvinar, fringilla lorem eget, ullamcorper urna.
          </p>
        </div>
      </div>

      <img
        class="relative block mx-auto mt-5 -mb-20 w-auto max-w-full"
        src="assets/img/introduction-music.png"
      />
    </section>

    <!-- Main Content -->
    <section class="container mx-auto">
      <div
        class="bg-white rounded border border-gray-200 relative flex flex-col"
      >
        <div
          v-icon-secondary="{ icon: 'headphones-alt', right: true }"
          class="px-6 pt-6 pb-5 font-bold border-b border-gray-200"
        >
          <span class="card-title">Songs</span>
        </div>

        <!-- Playlist -->
        <ol id="playlist">
          <SongItem v-for="song in songs" :song="song" :key="song.id" />
        </ol>
      </div>
    </section>
  </main>
</template>

<script lang="ts">
import { defineComponent, reactive, toRefs, onBeforeUnmount } from "vue";

import iconSecondary from "@/directives/iconSecondary";
import SongItem from "@/components/SongItem.vue";
import { songsCollection } from "@/includes/firebase";
import { Song, SongWithId } from "@/types/Song";

interface HomeState {
  songs: SongWithId[];
  numSongsPerPage: number;
  requestPending: boolean;
}

export default defineComponent({
  name: "Home",

  components: {
    SongItem,
  },

  setup() {
    const state = reactive<HomeState>({
      songs: [],
      numSongsPerPage: 25,
      requestPending: false,
    });

    const listSongs = async () => {
      if (state.requestPending) return; // Prevent triggering request multiple times in parallel, resulting in inconsist results
      state.requestPending = true;

      // Limit
      let query = songsCollection
        .orderBy("modifiedName")
        .limit(state.numSongsPerPage);

      // Offset
      if (state.songs.length) {
        const lastDoc = await songsCollection
          .doc(state.songs[state.songs.length - 1]?.id)
          .get();
        query = query.startAfter(lastDoc);
      }

      const songsSnapshot = await query.get();
      songsSnapshot.forEach((doc) =>
        state.songs.push({ id: doc.id, ...(doc.data() as Song) })
      );

      state.requestPending = false;
    };

    const handleScroll = () => {
      const { scrollTop, offsetHeight } = document.documentElement;
      const { innerHeight } = window;
      const isScrolledToBottom =
        Math.ceil(scrollTop) + innerHeight >= offsetHeight;
      if (!isScrolledToBottom) return;
      listSongs();
    };

    listSongs();
    window.addEventListener("scroll", handleScroll);
    onBeforeUnmount(() => {
      window.removeEventListener("scroll", handleScroll);
    });

    return toRefs(state);
  },

  // Register directives locally
  directives: {
    "icon-secondary": iconSecondary,
  },
});
</script>
