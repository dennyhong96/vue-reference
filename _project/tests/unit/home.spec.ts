import { shallowMount } from "@vue/test-utils";

import Home from "@/views/Home.vue";
import SongItem from "@/components/SongItem.vue";
import { TEST_SONGS } from "./__mocks__";
import { Howl } from "howler";

// Mock firebase module
jest.mock("@/includes/firebase", () => {
  //
});

describe("Home.vue", () => {
  test("should render a list of songs", async () => {
    // Overwrite component setup method, prevent network request, pass in data directly
    if (Home.setup) {
      Home.setup = () => ({ songs: TEST_SONGS });
    }

    const wrapper = shallowMount(Home, {
      data() {
        return { songs: TEST_SONGS };
      },

      global: {
        mocks: {
          // Mock $t globally
          $t(str: string) {
            return str;
          },
        },
      },
    });

    // wrapper.element is the root DOM element of the component
    expect(wrapper.element).toMatchSnapshot();

    const items = wrapper.findAllComponents(SongItem); // Select all by component

    // Render the correct number of songs
    expect(items).toHaveLength(TEST_SONGS.length);

    // Render songs in the correct order
    items.forEach((itemWrapper, idx) => {
      expect(itemWrapper.props().song).toEqual(TEST_SONGS[idx]);
    });
  });
});
