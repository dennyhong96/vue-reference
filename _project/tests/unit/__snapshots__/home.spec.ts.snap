// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue should render a list of songs 1`] = `
<main>
  <!-- Introduction -->
  <section
    class="mb-8 py-20 text-white text-center relative"
  >
    <div
      class="absolute inset-0 w-full h-full bg-contain introduction-bg"
      style="background-image: url(assets/img/header.png);"
    />
    <div
      class="container mx-auto"
    >
      <div
        class="text-white main-header-content"
      >
        <!-- $t() is available after registering i18n as a plugin -->
        <h1
          class="font-bold text-5xl mb-5"
        >
          home.listen
        </h1>
        <p
          class="w-full md:w-8/12 mx-auto"
        >
           Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus et dolor mollis, congue augue non, venenatis elit. Nunc justo eros, suscipit ac aliquet imperdiet, venenatis et sapien. Duis sed magna pulvinar, fringilla lorem eget, ullamcorper urna. 
        </p>
      </div>
    </div>
    <img
      class="relative block mx-auto mt-5 -mb-20 w-auto max-w-full"
      src="assets/img/introduction-music.png"
    />
  </section>
  <!-- Main Content -->
  <section
    class="container mx-auto"
  >
    <div
      class="bg-white rounded border border-gray-200 relative flex flex-col"
    >
      <div
        class="px-6 pt-6 pb-5 font-bold border-b border-gray-200"
      >
        <span
          class="card-title"
        >
          Songs
        </span>
        <i
          class="fa fa-headphones-alt text-xl text-green-400 float-right"
        />
      </div>
      <!-- Playlist -->
      <ol
        id="playlist"
      >
        
        <song-item-stub
          song="[object Object]"
        />
        <song-item-stub
          song="[object Object]"
        />
        <song-item-stub
          song="[object Object]"
        />
        
      </ol>
    </div>
  </section>
</main>
`;
